### ISSUE TITLE
High memory consumption in release prod

### ISSUE THREAD

**Issue by bpoplauschi** [rs/SDWebImage#2448]
### New Issue Checklist

* [X] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [X] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [X] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | e.g. 11.0 / 10.13.0 / 11.0 / 4.0
 SDWebImage Version      | 5.0.0-beta2
 Integration Method      | cocoapods
 Xcode Version           | e.g. Xcode 9 / Xcode 8
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Author: @fschaus

Hi guys,

I am having a very similar issue to #2418 (using GIFs with v5.0 Beta 2) but can't get hold of the of the shouldDecompress method/ variable in Swift (went through the doc but couldn't find it)...

Do you mind letting me know how to set off the decompression in Swift? I'll try to see if this works before opening a new issues (if necessary).

Thanks!
Francois

**Comment by bpoplauschi** [comment]
@dreampiggy responded:
@fschaus Hi. The SDWebImage 5.x, change that global control level `shouldDecompress`, to the `SDWebImageOptions` level, which allow you to control each image request instead of a global one (Because some small images, it's no need to trun off that force decode feature, which cause a frame rate drop). See the related PR for this #2283 #1927 

You can also find it on the migration guide here : [5.0-Migration-guide](https://github.com/rs/SDWebImage/wiki/5.0-Migration-guide). The correspond options, was moved from cache/loader-level , to each image request level.

You should now use like this to disable the force decode feature:

```swift
imageView.sd_setImage(with: url, placeholderImage: nil, options: [.avoidDecodeImage])
```

And also, could I ask the issue why cause your application out of memory ? Did you setup a maxmium cache size `SDImageCacheConfig.maxMemoryCost or maxMemoryCount`  to avoid cache too much of image in memory ? Did you try to load animation GIF on normal UIImageView ? (You can disable animation suppoert, by using `firstFrameOnly` options for each image request, or totally remove `SDImageGIFCoder` from coders list)

**Comment by bpoplauschi** [comment]
@fschaus responded:
Thanks for the input, I should have looked in the migration guide.

So I profiled a bunch of tests on my app to get to the bottom of this and see really weird behavior. See the link below for the details:
• My debug version runs as expected - once loaded my app allocates ~70Mb in memory (which is in line with what I saw previously when I was using KF). After a bit of usage (message loading + images) it goes up to 120-13Mb, rarely above (maxCache works) and flushes itself once the app enters background.
• My release/ prod scheme however shots straight to 160Mb at load, quickly goes to 300Mb (or more if I keep loading but then more often than not crashes) which doesn't respect my cache limit. If I don't turn off decoding the memory doesn't get flushed once the app enters the background (it decoding is off but still retains more than expected at 120-140Mb vs. 20-30Mb w/ my debug scheme).
• I use 5.0 beta 2, with a mix of fixed images and gifs (the initial loading screens only have fixed images). I also tested with UIImageView and AnimatedImgView but that makes no difference
• I also used Kingfisher to compare and its behavior with my prod release was similar to SDWI with my debug scheme. In all cases I use the "default" manager settings (except decode that I toggle on/ off for one of my use case).

My detailed tests are here: https://drive.google.com/open?id=1HIRHttQYn1NIJr0Omb9BG2FLOeWa8E_9

I am not deep enough in this library nor in iOS image frameworks to understand how they differ between the debug and the release scheme but I feel like the culprit is likely there (given the vastly different behavior observed).

I really like SDWebImg as it is the only library with a strong gif manager which allows me to have a single cache/ image processor for PGN & Gifs (I had a weird convoluted kingfisher + FLAnimatedImg setup before) and like how more flexbile SDWebImg is (e.g., add fade, progress bar, etc.) so really hope we can get to the bottom of this together J

Thanks!
Francois

**Comment by dreampiggy** [comment]
@SamboVisal Hi. Thanks for the detailed report. So that I can start to investigate this issue, and the fix may be available in soon (Maybe another 5.0.0-beta3 ?)

After reading you report, I have some other detailed information to know, so could you please also reply me about it, to make it easier for us to find out the reason.

+ Are the test about `Animated ImageView` represent `SDAnimatedImageView` ? Or the `FLAnimatedImageView` ?
+ Are the following test running on iPhone read device, instead of iPhone Simulator ? (iPhone Simulator has some issue for the `UIApplicationDidReceiveMemoryWarningNotification`, because of the simulator is running on host Mac env, and it can only be triigered once)
+ Could you please also test the 5.x branch instead of 5.0.0-beta release ? (Use `pod 'SDWebImage', :git => 'https://github.com/rs/SDWebImage.git', :branch => '5.x'`), and there are one options in `SDImageCacheConfig.shouldUseWeakMemoryCache`, could you also try to turn it off ? This options is available in SD 4.x, but it's broken on 5.0.0-beta2 release, we fix it on the latest 5.x branch)

**Comment by dreampiggy** [comment]
@fschaus Hi...Today I try to upgrade one Social App in our company, to the latest 5.x branch. We also have a feed list, but there are only 10 images per screen.

I test it with Profile (Release Mode) && Debug (Debug Mode), the memory usage is nearly the same, and when enterging the background, the memeory usage was also dropped.

So, did you try the latest 5.x branch ? Or could you please tell us if there are some extra config for you to use, like anything in the `SDImageCacheConfig` ?

**Comment by fschaus** [comment]
Fixed! Well done and thanks for that!

The memory usage now kicks in at 50Mb or so and gradually/ slowly increases as more pictures get displayed/ cached. Bringing the app to the background nicely flushes everything. I tried with both UIImage and SDAnimatedImage and both give me the right profile.

The only thing that continues to go above my cache limit (100Mb) is when I display GIFs - it easily brings the memory to 200-300Mb but I suspect that's the decoding vs. the cache. 

Great job!
Francois

**Comment by dreampiggy** [comment]
@fschaus Hi.
Did you means that the SDWebImage 5.x branch fix the issue ? Or only if you setup some extra config, like `SDImageCacheConfig.shouldUseWeakMemoryCache` or some other, cause this issue go away ?

Another tip: About `SDAnimatedImageView`, the rendering memory usage will not follow the SDImageCache.maxMemorySize, because we check the memory cache size, by using `image.size.height * image.size.width * image.scale * image.scale`. However, that `SDAnimatedImage` use a buffer cache of image frames. And by default, it was calculated by considering the current free memory and total memory, so it's normally OK and may consume some  memory with a `safe` limit. And you can also setup your own value for each image view.

See [Animated Image Rendering-maxBufferSize](https://github.com/rs/SDWebImage/wiki/Advanced-Usage#rendering).

**Comment by fschaus** [comment]
The new commit fixed the issue! I didn’t make any other changes - just used the new commit and the default settings.

Thanks again!

**Comment by dreampiggy** [comment]
So now we can close this issue. Any extra consideration about this ?
